### PR TITLE
PLT-8079 - Added functionality to take into account a possible upperSlot when calculating stability

### PR DESCRIPTION
**Pull request by willjgould**
* Support for `upperSlotNo` when querying with stability
* Refactor of the stability calculation function to make it more modular (it's essentially just about error handling now)
* A new indexer type, `MintTokenEventIndexerQuery`, containing of a `MintTokenEventIndexer` and a `BlockInfoIndexer`
* Testing machinery for `MintTokenEventIndexerQuery`, plus tests

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting main unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [x] Reviewer requested

### CODE REVIEW

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/MintTokenEvent.hs [hunk 1]
```diff
@@ -726,15 +745,94 @@ instance
 
 instance
   ( MonadIO m
+  , MonadError (Core.QueryError (QueryByAssetId MintTokenBlockEvents)) m
+  )
+  => Core.IsSync m event MintTokenEventIndexerCombine
+  where
+  lastSyncPoint idx = do
```
Comment (willjgould) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/MintTokenEvent.hs:752]: Should this take the `lastSyncPoint`s from both indexers and take the minimum instead?

**Review by koslambrou — CHANGES_REQUESTED** [review]

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/MintTokenEvent.hs [hunk 2]
```diff
@@ -120,13 +128,22 @@ import Marconi.ChainIndex.Indexers.Worker (
...
 
 -- | A raw SQLite indexer for 'MintTokenBlockEvents'
 type MintTokenEventIndexer = Core.SQLiteIndexer MintTokenBlockEvents
 
+-- | The inner state of the 'MintTokenIndexerCombine'
+data MintTokenEventIndexerCombine event = MintTokenEventIndexerCombine
+  { _mintTokenEventIndexerSecurityParam :: !SecurityParam
+  , _mintTokenEventIndexerCombineIndexer :: Con.MVar (StandardIndexer IO Core.SQLiteIndexer event)
+  , _mintTokenEventIndexerCombineBlockInfo :: Con.MVar (StandardIndexer IO Core.SQLiteIndexer BlockInfo)
+  }
```
[the remark this answers is not in the document]
Comment (willjgould) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/MintTokenEvent.hs:145]: I did consider it, but I didn't think it was necessary. However, I also don't fully know the benefits of using it over how I've done it. I presumed it existed to allow us to do more complex queries between tables, and given my queries are quite simple (we use the same `BlockInfo` result for all `Stability` calculations) I figured it was unnecessary complexity. Is there some other benefit that I've missed?

[the remark this answers is not in the document]
Comment (willjgould) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/MintTokenEvent.hs:145]: I'll move it to a new module as is, as discussed on a call.

Will add a comment stating the possibility of moving to the AggregateQuery.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/MintTokenEvent.hs [hunk 3]
```diff
@@ -123,6 +144,25 @@ tests =
             "propRunnerDoesntTrackUnselectedAssetId"
             propRunnerDoesntTrackUnselectedAssetId
         ]
+    , testGroup
+        "Stability"
+        [ testPropertyNamed
+            "withStability always stable when SecurityParam is 0"
+            "withStabilityAllStable"
+            withStabilityAllStable
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/MintTokenEvent.hs:152]: Small convention. Property tests should start with `prop` such as `propWithStabilityAllStable`.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/MintTokenEvent.hs [hunk 4]
```diff
@@ -699,3 +739,248 @@ genTrackedAssetIds assetIds = do
...
+        withIndexer securityParam events $
+          Core.query
+            point
+            ( Core.WithStability
+                (MintTokenEvent.QueryByAssetId policy Nothing Nothing usn Nothing)
+            )
+  assertion res
+
+-- | Generate a list of events from a mock chain
+getMintTokenEvents
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/MintTokenEvent.hs:810]: getMintTokenEvents -> genMintTokenEvents

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/MintTokenEvent.hs [hunk 5]
```diff
@@ -699,3 +739,248 @@ genTrackedAssetIds assetIds = do
...
+        events <- concat <$> traverse (getInputs blockNo . getTxBody) txs
+        pure $
+          Core.Timed
+            (C.ChainPoint slotNo blockHeaderHash)
+            ( fmap MintTokenBlockEvents $
+                NonEmpty.nonEmpty events
+            )
+   in traverse getBlockSpentsEvent
+
+getInputs :: C.BlockNo -> C.TxBody era -> Gen [MintTokenEvent]
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/MintTokenEvent.hs:829]: getInputs -> genInputs

**Review by koslambrou — APPROVED** [review]
LGTM. Small fixes.

**Pull request merged**